jetty-io/src/main/java/org/eclipse/jetty/io/AbstractConnection.java

```diff
@@ -311,24 +323,40 @@ public String toConnectionString()
             this);
     }
 
-    private class ReadCallback implements Callback
+    private class ReadCallback implements Callback, Invocable
     {
         @Override
         public void succeeded()
         {
+            if (isUseVirtualThreadToInvokeRootHandler())
+            {
+                if (VirtualThreads.startVirtualThread(AbstractConnection.this::onFillable))
+                    return;
+            }
             onFillable();
         }
 
         @Override
-        public void failed(final Throwable x)
+        public void failed(Throwable x)
         {
+            if (isUseVirtualThreadToInvokeRootHandler())
+            {
+                if (VirtualThreads.startVirtualThread(() -> onFillInterestedFailed(x)))
+                    return;
+            }
             onFillInterestedFailed(x);
         }
 
+        @Override
+        public InvocationType getInvocationType()
+        {
+            return isUseVirtualThreadToInvokeRootHandler() ? InvocationType.NON_BLOCKING : InvocationType.BLOCKING;
```

Comment:
It feels like this logic should be elsewhere as well.... ie for the initial dispatch to the handler?